presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -304,16 +305,18 @@ else if (partitioningHandle.equals(SCALED_WRITER_DISTRIBUTION)) {
                     connectorPartitionHandles = ImmutableList.of(NOT_PARTITIONED);
                 }
 
+                ConnectorBucketNodeMap connectorBucketNodeMap = nodePartitioningManager.getConnectorBucketNodeMap(session, partitioningHandle);
+
                 BucketNodeMap bucketNodeMap;
                 List<Node> stageNodeList;
                 if (plan.getSubStages().isEmpty() // no remote source
                         && groupedExecutionForStage
                         // connector has no fixed mapping
-                        && !nodePartitioningManager.getConnectorBucketNodeMap(session, partitioningHandle).hasFixedMapping()) {
+                        && !connectorBucketNodeMap.hasFixedMapping()) {
                     // dynamic bucket schedule
                     stageNodeList = new ArrayList<>(nodeScheduler.createNodeSelector(connectorId).allNodes());
                     Collections.shuffle(stageNodeList);
-                    bucketNodeMap = nodePartitioningManager.getDynamicBucketNodeMap(session, partitioningHandle);
+                    bucketNodeMap = nodePartitioningManager.getDynamicBucketNodeMap(connectorBucketNodeMap);
```

Comment:
I don't think it's the right abstraction for `NodePartitioningManager` to expose `getConnectorBucketNodeMap`. And I think it is the right abstraction for `getDynamicBucketNodeMap` to take `session` and `partitioningHandle`, just like `getNodePartitioningMap` does.

Problem is, you need to do `hasFixedMapping`. For this, you can pass a `dynamicEligible` or `preferDynamic` when you invoke the method in `nodePartitioningManager`. If you do this, the next question will be the return type of the said method. It will need to return either `NodePartitionMap` or `DynamicBucketNodeMap`. I think it could be reasonable to return an object that has 3 fields: `isDynamic`, `nodePartitionMap`, `dynamicBucketNodeMap`.

I think the approach taken by this commit is not in the right direction. Ideally, `NodePartitionMap` should not expose a public factory method for `ConnectorBucketNodeMap`. This commit doubles down on that by requiring user of `NodePartitionMap` to pass the created object back into `NodePartitionMap`.

Reply:
Per discussion, when `plan.getSubStages().isEmpty()` holds, it's anyway a colocated join (includes replicated join in the future), and `NodePartitionMap` is not required in this case.

Thus, we can make `nodePartitioningManager` to take `dynamicEligible`. And returns `FixedBucketNodeMap` / `DynamicBucketNodeMap` based on whether connector runs in shared-nothing or shared-storage mode.